enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+
+![Egress IP VRF LGW](images/egress-ip-vrf-lgw.svg)
+
+
+
+#### Egress Firewall
+
+Egress firewall is enforced at the OVN logical switch, and this proposal has no effect on its functionality.
+
+
+#### Egress QoS
+
+Egress QoS is namespace scoped and functions by marking packets at the OVN logical switch, and this proposal has no effect on its functionality.
+
+
+#### Egress Service
+
+Egress service functions similar to Egress IP in local gateway mode, with the exception that all traffic paths go through the kernel networking stack. Egress Service also uses IP rules and VRFs in order to match on traffic and forward it out the right network (if specified in the CRD). It uses iptables in order to SNAT packets to the load balancer IP. Like Egress IP, with unique networks there will need to be IP rules with higher precedence to match on packets from specific networks and direct them to the right VRF.
+
+
+#### Multiple External Gateways (MEG)
+
+There will be no support for MEG or pod direct ingress on any network other than the primary default network. Remember, MEG works the same way in local or shared gateway mode, by utilizing the shared gateway path. This support may be enhanced later by extending VRFs/networks outside of the cluster.
+
+
+### Host Networked Pods
+
+By encompassing VRFs into the host, this introduces some constraints and requirements for the behavior of host networked type pods. If a host networked pod is created in a Kubernetes namespace that has a unique network, should it be restricted to only communicating with that Kubernetes network? If the answer is yes, then we would need to do the following:
+
+With Linux VRFs, the application would need to bind directly to the VRF device in order to use TCP/UDP type sockets on the network. Raw sockets by default are allowed to span multiple VRFs. This seems like a problematic solution, as we cannot expect all user applications to provide configuration for which device to bind to for TCP/UDP sockets. There is sysctl configuration that will allow TCP/UDP sockets to listen across all VRFs, but that means then a host application could communicate with any network:
+
+
+```
+       sysctl -w net.ipv4.tcp_l3mdev_accept=1
+       sysctl -w net.ipv4.udp_l3mdev_accept=1
+```
+
+
+In order to restrict applications from being able to talk on the other networks, it seems like the only choice would be to insert iptables rules to block traffic coming from/to any of the network interfaces for that application. See [https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/networking/vrf.rst?h=v6.1](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/networking/vrf.rst?h=v6.1) for more details.
```

Comment:
> In order to restrict applications from being able to talk on the other networks

HostNetwork pods from different networks could still communicate with each other across `127.0.0.1`, right?

Reply:
and wouldn't we only be blocking layer >= 3 traffic ?... L2 would still be available.

Reply:
The loopback interface is bound to the default VRF. So yeah if the host network pods are not bound to the VRF device then they can communicate over the loopback or via any other IP in the default VRF. However, we can block them from being able to communicate with ovn-networked pods in the user-defined network. In the future, if we can force the bind call with ebpf to ensure it binds to the right VRF we can plug this hole, but for now it seems like it isnt a huge security problem. I can add some context around this as well in the document. Thanks for the input.

@maiqueb I think it is all layer 3 when coming from the host networked pod. Since we do not attach mp0 to a linux bridge, and then force the application to bind to the bridge there cannot be raw layer 2 communication. So every packet that is sent by the host networked pod will be routed by the kernel into mp0, implying IP (even though the OVN topology is all layer 2). Do you think this is an issue for users?